### PR TITLE
Speedup for molecule isomorphism checking

### LINKED ISSUE #2035: `Topology.identical_molecule_groups` is too slow
**Describe the bug**

Subgraph isormorphism is central to using the toolkit on multi-molecule systems but it is slow, especially for large and complicated systems.

**To Reproduce**

[topology.json.zip](https://github.com/user-attachments/files/19322417/topology.json.zip)

With a topology of fairly modest size and refactoring to use a Rust re-implementation of `networkx` (#2033), this takes about 10 minutes:

```python
In [1]: from openff.toolkit import Topology

In [2]: topology = Topology.from_json(open("topology.json").read())

In [3]: %timeit -o -r1 -n1 topology.identical_molecule_groups
11min 2s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
Out[3]: <TimeitResult : 11min 2s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

In [4]: !open .

In [5]: topology.n_atoms, topology.n_molecules
Out[5]: (4119, 10)
```

Using the current `main` branch, it takes at least twice that (24 minutes, still running):

Image

**Output**

**Computing environment (please complete the following information):**
 - Operating system
 - Output of running `conda list`

**Additional context**

#1143 #1734 #353 #2008 https://github.com/openforcefield/openff-interchange/issues/1156 etc.

### PR DESCRIPTION
**Pull request by j-wags**
- [x] Fix #2035
    - This makes the reproducing case run on my machine in ~0.5 sec, compared to 11 min using the old stack 
- [x] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/main/openff/toolkit/_tests)
- [x] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/main/docs), if applicable
- [x] [Lint](https://docs.openforcefield.org/projects/toolkit/en/stable/users/developing.html#style-guide) codebase
- [x] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/main/docs/releasehistory.md)

### CODE REVIEW

File: openff/toolkit/topology/molecule.py [hunk 1]
```diff
@@ -2151,6 +2151,12 @@ def to_networkx(data: FrozenMolecule | nx.Graph) -> nx.Graph:
                 return data.to_networkx()
 
             elif isinstance(data, nx.Graph):
+                for node in data:
+                    h_counter = -1
+                    for neighbor in data.neighbors(node):
+                        if data.nodes[neighbor]['atomic_number'] == 1:
+                            data.nodes[neighbor]['atomic_number'] = h_counter
+                            h_counter -= 1
```
Comment (mattwthompson) [inline openff/toolkit/topology/molecule.py]: Does this depend on the order of the `data.neighbors(node)` iterator (in problematic ways)? I will try to think of an antagonistic test case tomorrow, if possible

**Comment by mattwthompson** [comment]
So far I'm seeing a net improvement in some `Interchange.to_x` export times. I have **not** looked at anything beforehand (molecule loading, topology preparation, parameter assignment, partial charge assignment, etc.) nor is this a completely comprehensive analysis.[^2]

The two environments I used are my Interchange development environment (with this branch installed) and a second with only released versions of our software[^1]

|         | OpenFE JACS set (vacuum) | Mixed solvent system | Ligand in water | Tim's polymers   |
|---------|-----------------|----------------------|-----------------|------------------|
| GROMACS | No change       | Marginally worse     | Slightly better | Huge improvement |
| OpenMM  | No change       | Marginally worse     | Somewhat better | No change        |


[^1]: `micromamba create --name released openff-toolkit seaborn polars tqdm matplotlib`
[^2]: I have receipts for everything elsewhere if people want lots of logs files

**Comment by j-wags** [comment]
A quick update here - I'm offline for the rest of this week and half of the next, but my plan is to come back around the 27th and try to prove to myself that this is a safe approach and put in tests to raise the alarm if a later change violates some assumption that this needs (no divalent H, no H isotopes, no negative atomic numbers, etc). If folks want to help out by thinking of edge cases/regressions, that could help me get through the QA process faster.
